packages/kotti-ui/source/props.test.ts

```diff
@@ -0,0 +1,451 @@
+			prop: z.function().default(() => undefined),
+		})
+
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, undefined)
+		expect(prop.type).toBe(Function)
+		expectValidator(prop, [() => undefined], ['string', null])
+
+		/**
+		 * HACK: This is actually a minor bug in zod, theoretically this
+		 * should be true, however `.function().default()` doesn’t set
+		 * this to optional correctly. This can be safely ignored
+		 * since vue doesn’t even run the validator when passing `undefined`
+		 *
+		 * @see {@link https://github.com/colinhacks/zod/issues/647}
+		 */
+		expectValidator(prop, [], [undefined])
+	})
+
+	it('z.function().nullable().default()', () => {
+		const schema = z.object({
+			prop: z.function().nullable().default(null),
+		})
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, null)
+		expect(prop.type).toBe(Function)
+		expectValidator(prop, [() => undefined, null, undefined], ['string', true])
+	})
+})
+
+describe('nativeEnum', () => {
+	enum TestEnum {
+		KEY_A = 'VALUE_A',
+		KEY_B = 'VALUE_B',
+		KEY_C = 'VALUE_C',
+	}
+
+	beforeAll(silenceConsole)
+
+	it('z.nativeEnum()', () => {
+		const schema = z.object({
+			prop: z.nativeEnum(TestEnum),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(String)
+		expectValidator(
+			prop,
+			[TestEnum.KEY_A, TestEnum.KEY_B, TestEnum.KEY_C],
+			['VALUE_D', 'string', false, [], null, undefined],
+		)
+	})
+
+	it('z.nativeEnum().nullable()', () => {
+		const schema = z.object({
+			prop: z.nativeEnum(TestEnum).nullable(),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(String)
+		expectValidator(
+			prop,
+			[TestEnum.KEY_A, TestEnum.KEY_B, TestEnum.KEY_C, null],
+			['VALUE_D', 'string', false, [], undefined],
+		)
+	})
+
+	it('z.nativeEnum().default()', () => {
+		const schema = z.object({
+			prop: z.nativeEnum(TestEnum).default(() => TestEnum.KEY_A),
+		})
+
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, TestEnum.KEY_A)
+		expect(prop.type).toBe(String)
+		expectValidator(
+			prop,
+			[TestEnum.KEY_A, TestEnum.KEY_B, TestEnum.KEY_C, undefined],
+			['VALUE_D', 'string', false, [], null],
+		)
+	})
+
+	it('z.nativeEnum().nullable().default()', () => {
+		const schema = z.object({
+			prop: z.nativeEnum(TestEnum).nullable().default(null),
+		})
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, null)
+		expect(prop.type).toBe(String)
+		expectValidator(
+			prop,
+			[TestEnum.KEY_A, TestEnum.KEY_B, TestEnum.KEY_C, null, undefined],
+			['VALUE_D', 'string', false, []],
+		)
+	})
+})
+
+describe('number', () => {
+	beforeAll(silenceConsole)
+
+	it('z.number()', () => {
+		const schema = z.object({
+			prop: z.number(),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Number)
+		expectValidator(
+			prop,
+			[1, 0],
+			[false, [], 'string', Number.NaN, null, undefined],
```

Comment:
so in typescript `Number.NaN` extends the `Number` interface

in JS, the `typeof` operation returns `typeof NaN` and `typeof Number.NaN` to be `number`

so I don't find this intuitively to be a failure scenario

feel free to point it out in a separate use-case just to point out that zod doesn't support it to be `number`

maybe also try other edge cases - from the `Integer` class like MAX_SAFE_INTEGER
or not

up to you

Reply:
edge cases can be more than the `SAFE_INTEGER`s ofc

Reply:
Regarding `NaN`, I’m pretty sure most developers aren’t even fully aware of it, especially not aware enough to always treat all their number handling with `NaN` in mind or to actively _expect_ `NaN` to be allowed to be passed to something that needs a number.

I think it’s a very sane default behavior of `zod` to exclude `NaN`. Even the `isNumber` function in the `KtFieldNumber` utilities explicitly forbids `NaN`.

However, I discovered two edge-cases that I actually did **not** expect to behave like they do in `zod`:

1. `Number.POSITIVE_INFINITY` (is valid)
2. `Number.NEGATIVE_INFINITY` (is valid)